Add gsheets source

sherifnada opened the pull request:
## What
Main change: 
* add google sheets source

Supporting change: 
* refactor python model generation to be re-usable by multiple projects, and use it to generate models for the google sheets API. 

## Checklist
- [x] *Finish integrating with the new source test suite*

## Recommended reading order

Main change: 
1. `google_sheets_source.py`
1. `helpers.py`
1. remaining files in the `source-google-sheets` directory

Supporting change: 
1. `generate-python-models.gradle`
1. `base-python/build.gradle`
1. everything else

sherifnada commented:
got into a fistfight with Git while doing this

sherifnada commented on README.md:
```diff
@@ -51,6 +51,7 @@ We love contributions to Airbyte, big or small.
 
 See our [Contributing guide](https://docs.airbyte.io/contributing/contributing-to-airbyte) on how to get started. Not sure where to start? We’ve listed some [good first issues](https://github.com/airbytehq/airbyte/labels/good%20first%20issue) to start with. You can also [book a free, no-pressure pairing session](https://drift.me/micheltricot/meeting) with one of our core contributors.
 
+
```
```suggestion
```

cgardens commented on airbyte-integrations/connector-templates/python-source/.dockerignore:
```diff
@@ -1 +1,3 @@
 build
+secrets
```
what secrets are needed in python source? also why do we have a different construction here for the .dockerignore than elsewhere? is there a reason we don't do like we do elsewhere? e.g.

```
* 
!python_source
```

cgardens commented on airbyte-integrations/connectors/source-google-sheets/.gitignore:
```diff
@@ -0,0 +1 @@
+secrets
```
is this not handled by the .gitignore at the top of the prooject: https://github.com/airbytehq/airbyte/blob/master/.gitignore#L8

cgardens commented on airbyte-integrations/connectors/source-google-sheets/google_sheets_source/google_sheets_source.py:
```diff
@@ -0,0 +1,128 @@
+            spreadsheet_metadata = Spreadsheet.parse_obj(client.get(spreadsheetId=spreadsheet_id, includeGridData=False).execute())
+            sheet_names = [sheet.properties.title for sheet in spreadsheet_metadata.sheets]
+            streams = []
+            for sheet_name in sheet_names:
```
to make sure i understand this is for a single google sheet the individual pages inside of it?

sherifnada replied:
correct. The terminology is Spreadsheet for the "entire sheet", and subsheets are called a Sheet

cgardens commented on airbyte-integrations/connectors/source-google-sheets/google_sheets_source/models/definitions/spreadsheet.yaml:
```diff
@@ -0,0 +1,65 @@
+---
+"$schema": http://json-schema.org/draft-07/schema#
```
is there a reason it's particularly helpful to define these as json schema as opposed to using pojos (or whatever the python equivalent is)? i'm realizing i haven't been as thoughtful as i should have about this in the past, but it seems like if we are not planning to serialize these structs (which i don't think we are), adding the json schema layer is more trouble than its worth and might bring in some incidental complexity. if there's some other benefit you're getting then go for it.

sherifnada replied:
switched this to use only PyaDantic models. So basically it's like autogenerated code without the autogeneration. We'd need to do it manually. Couldn't find many great ways to do this otherwise.

cgardens commented on airbyte-integrations/connectors/source-google-sheets/integration_tests/before.py:
```diff
@@ -0,0 +1,77 @@
+SCOPES = ['https://www.googleapis.com/auth/spreadsheets', 'https://www.googleapis.com/auth/drive', 'https://www.googleapis.com/auth/drive.file']
+SECRETS_PATH = "../secrets/config.json"  # TODO replace with injected secrets
+
+# TODO this whole file is one big TODO. placeholder code for pre and post actions when running integration tests
```
can we get by without this and just host a single standard google sheet that all tests are run against?

cgardens replied:
same for after.

cgardens replied:
discussed offline. we prefer having before / after. sherif is going to take a look at how much work it is to add it to standard test source and make a judgmet call. we will add this feature to the standard test suite soon either way. just a question of whether it will happen before or after this pr.

cgardens commented on airbyte-integrations/connectors/source-google-sheets/google_sheets_source/google_sheets_source.py:
```diff
@@ -0,0 +1,128 @@
+            column_index_to_name = sheet_to_column_index_to_name[sheet]
+            row_cursor = 2  # we start syncing past the header row
+            encountered_blank_row = False
+            while not encountered_blank_row:
```
so it's not supported to have a blank row in the middle of your data? i guess that's fine.

cgardens commented on airbyte-integrations/connectors/source-google-sheets/google_sheets_source/google_sheets_source.py:
```diff
@@ -0,0 +1,128 @@
+        logger.info(f"Starting syncing spreadsheet {spreadsheet_id}")
+
+        sheet_to_column_index_to_name = Helpers.get_available_sheets_to_column_index_to_name(client, spreadsheet_id, sheet_to_column_name)
+        for sheet in sheet_to_column_index_to_name.keys():
```
would you mind adding some comments here to help understand what's going on. if i understand correctly it's:
```
for each page in google sheet
  fetch batch of rows
    for each row in batch emit
```

cgardens commented on airbyte-integrations/connectors/source-google-sheets/google_sheets_source/helpers.py:
```diff
@@ -0,0 +1,168 @@
+        return service_account.Credentials.from_service_account_info(creds_json, scopes=scopes)
+
+    @staticmethod
+    def headers_to_airbyte_stream(sheet_name, header_row_values: List[str]) -> AirbyteStream:
```
for all of these converters helpers, would you mind writing helpers?

cgardens commented on airbyte-integrations/connectors/source-google-sheets/google_sheets_source/helpers.py:
```diff
@@ -0,0 +1,168 @@
+        return sheet_to_column_name
+
+    @staticmethod
+    def row_data_to_record_message(sheet_name: str, cell_values: Values, column_index_to_name: Dict[int, str]) -> AirbyteRecordMessage:
```
same for unit testing

cgardens commented on airbyte-integrations/connectors/source-google-sheets/google_sheets_source/google_sheets_source.py:
```diff
@@ -0,0 +1,128 @@
+
+        return AirbyteConnectionStatus(status=Status.SUCCEEDED)
+
+    def discover(self, logger: AirbyteLogger, config_container) -> AirbyteCatalog:
```
is there a unit test that could be written for this logic (or some subset of it). alot of transformations going on.

cgardens commented on airbyte-integrations/connectors/source-google-sheets/google_sheets_source/google_sheets_source.py:
```diff
@@ -0,0 +1,128 @@
+                reason = "Requested spreadsheet was not found."
+            raise Exception(f"Could not run discovery: {reason}")
+
+    def read(self, logger: AirbyteLogger, config_container, catalog_path, state=None) -> Generator[AirbyteMessage, None, None]:
```
same as discover. is there a unit test that could be written for this logic (or some subset of it). alot of transformations going on.

cgardens reviewed (approved):
Looks good. Some comments. I think some unit tests would be clutch since there's a lot of data transformation happening. Also a question about the integration tests. Will reach out offline on that one.

michel-tricot commented on airbyte-integrations/connectors/source-google-sheets/setup.py:
```diff
@@ -0,0 +1,44 @@
+    author_email='contact@airbyte.io',
+    packages=['google_sheets_source'],
+    package_data={
+        '': ['models/generated/*.py']
```
py are not package data, they will get automatically packaged as it is considered code

jrhizor commented on airbyte-integrations/connectors/source-google-sheets/Dockerfile:
```diff
@@ -0,0 +1,15 @@
+FROM airbyte/integration-base-python:dev
+
+RUN apt-get update && apt-get install -y jq curl bash
```
```suggestion
RUN apt-get update && apt-get install -y jq curl bash && rm -rf /var/lib/apt/lists/*
```

Can do this like other integrations to reduce the image size.

jrhizor commented on airbyte-integrations/connectors/source-google-sheets/google_sheets_source/__init__.py:
```diff
@@ -0,0 +1,26 @@
+SOFTWARE.
+"""
+
+from .google_sheets_source import *
```
should specify each import and use them in an `__all`

sherifnada replied:
done in local

jrhizor commented on tools/gradle/commons/generate-python-models.gradle:
```diff
@@ -0,0 +1,17 @@
+    }
+
+    outputs.upToDateWhen { false }
+    dependsOn ':airbyte-integrations:bases:base-python:code-generator:buildImage'
```
This seems like it should be passed in from `ext`?

jrhizor commented on build.gradle:
```diff
@@ -82,7 +82,9 @@ spotless {
 }
 check.dependsOn 'spotlessApply'
-spotlessPython.dependsOn ':airbyte-integrations:bases:base-python:generateProtocolClassFiles'
+spotlessPython.dependsOn ':airbyte-integrations:bases:base-python:generatePythonModels'
```
These should probably live in their integration's build files if possible.

jrhizor reviewed (commented):
Can your `README.md` link to https://docs.airbyte.io/integrations/sources/googlesheets ?

Also, can those docs be added as part of this PR?

sherifnada commented:
@michel-tricot @cgardens @jrhizor @ChristopheDuong responded to all comments I think. changes since last review: 
1. Added unit tests
1. Added integration tests following the standard source tests + added them to CI
1. Removed autogen of models and now instead using handcoded PyDantic models. 
1. Added README to package

cgardens commented on airbyte-integrations/connectors/source-google-sheets/integration_tests/integration_test.py:
```diff
@@ -0,0 +1,121 @@
+]
+
+
+class GoogleSheetsSourceStandardTest(StandardSourceTestIface):
```
i like this alot. can you add issues for other features that would have made this easier? seems like the main thing is not having to write files would have been nice in your case.

cgardens commented on airbyte-integrations/connectors/source-google-sheets/main_dev.py:
```diff
@@ -0,0 +1,33 @@
+
+import sys
+
+from airbyte_protocol.entrypoint import launch
```
```suggestion
from base_python.entrypoint import launch
```

cgardens commented on airbyte-integrations/connectors/source-google-sheets/setup.py:
```diff
@@ -0,0 +1,48 @@
+    author="Airbyte",
+    author_email="contact@airbyte.io",
+    packages=find_packages(),
+    install_requires=[
```
is using install_requires intentional here? it means your standard tests are going to depend on what's on line 46 AND everything in this array as well. if it's intentional that's fine. just wanted to make sure it was clear.

sherifnada replied:
it's intentional since the tests pretty much depend on the same things as the main package

cgardens commented on airbyte-integrations/connectors/source-google-sheets/setup.py:
```diff
@@ -0,0 +1,48 @@
+        "requests",
+        "google-api-python-client",
+        "google-auth-httplib2",
+        "airbyte-protocol",
```
```suggestion
        "airbyte-protocol",
        "base-python",
```

cgardens commented on airbyte-integrations/connectors/source-google-sheets/unit_tests/test_helpers.py:
```diff
@@ -0,0 +1,196 @@
+from google_sheets_source.models import CellData, GridData, RowData, Sheet, SheetProperties, Spreadsheet
+
+
+class TestHelpers(unittest.TestCase):
```
why is this TestHelpers? aren't these the tests themselves?

cgardens replied:
oops. these are tests for helpers. my bad. your name is fine.

cgardens commented on tools/bin/standard_test_pr.sh:
```diff
@@ -7,4 +7,5 @@ set -e
 ./gradlew --no-daemon --scan -x generateProtocolClassFiles \
-  :airbyte-integrations:connectors:source-github-singer:standardSourceTestPython
+  :airbyte-integrations:connectors:source-github-singer:standardSourceTestPython \
+  :airbyte-integrations:connectors:source-google-sheets:standardSourceTestPython
```
any ratelimiting on the google sheets API that we should be worried about?

sherifnada replied:
nothing to be worried about, it's 500 requests per 100 seconds. If we have problems with this, it'll mean our build is blazing fast ;)

cgardens commented on airbyte-integrations/connectors/source-google-sheets/README.md:
```diff
@@ -0,0 +1,38 @@
+
+## Configure credentials
+### Configuring credentials as a community contributor
+Follow the instructions in the documentation above for generating credentials to access the Google API, then put those 
```
these instructions are where? gitbook? if thery are elsewhere that's fine. just pointing out that in this doc there doesn't appear to be any information on how to get your google api creds.

sherifnada replied:
fixed, these are below

sherifnada replied:
oops I misunderstood what this was referring to. This is in Gitbook, will update the link.

cgardens reviewed (approved):
i dig it. nice work! our first full python source. some small questions and comments, but do not block merging on them.

sherifnada commented on airbyte-integrations/connectors/source-google-sheets/README.md:
```diff
@@ -0,0 +1,38 @@
+
+## Configure credentials
+### Configuring credentials as a community contributor
+Follow the instructions in the documentation above for generating credentials to access the Google API, then put those 
```
```suggestion
Follow the instructions below for generating credentials to access the Google API, then put those 
```
